configurable ARCHIVE_FILE_SIZE_LIMIT

hiroyuki-sato opened the pull request:
This PR introduce new `api.max_archive_total_size_limit` parameter for fixing #830


## Before the fix

```
error: Status code 400: {"message":"Size of the uploaded archive file exceeds limit (2097152 bytes)","status":400}
```


### After the fix

digdag.conf

```
api.max_archive_total_size_limit = 6291456 # 6MB 1024 * 1024 * 6
```

```
./pkg/digdag-0.9.33.jar server -m -c /tmp/digdag.conf
```

```
digdag push fuga
2019-02-18 15:28:50 +0900: Digdag v0.9.32
Creating .digdag/tmp/archive-2607659884924645158.tar.gz...
  Archiving gawk-4.2.1.tar.xz
  Archiving fuga.dig
  Archiving scripts/myscript.sh
  Archiving test.sh
  Archiving 2m
...

Use `digdag workflows` to show all workflows.
```

hiroyuki-sato commented:
@sonots Thank you for reviewing this PR. 
I'll add tests later. I don't know how to write test code yet.

hiroyuki-sato commented:
Hello, @yoyama and @muga 

Could you point me how to test this PR?
There is no tests `io.digdag.server.rs` package. 

https://github.com/treasure-data/digdag/tree/master/digdag-server/src/test/java/io/digdag/server

sonots-zozo commented:
Hi, I want this feature as soon as possible. Waiting to be merged and released.

hiroyuki-sato commented:
Add note.

Although without test code, I confirmed with `api.max_archive_total_size_limit` config and default behavior(without  `api.max_archive_total_size_limit`) by hand.
This PR doesn't break the current implementation.

muga commented on digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java:
```diff
@@ -157,8 +157,10 @@
-    private static final int ARCHIVE_TOTAL_SIZE_LIMIT = 2 * 1024 * 1024;
-    private static final int ARCHIVE_FILE_SIZE_LIMIT = ARCHIVE_TOTAL_SIZE_LIMIT;
+    private static final int DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT = 2 * 1024 * 1024;
+    private static final int DEFAULT_ARCHIVE_FILE_SIZE_LIMIT = DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT;
```
Please tell me why `DEFAULT_ARCHIVE_FILE_SIZE_LIMIT` is necessary?

hiroyuki-sato replied:
No reason. It based on @sonots [comment](https://github.com/treasure-data/digdag/pull/994#commitcomment-32353677). but he also doesn't have any reason. Shall I remove it?

sonots-zozo replied:
There are two ARCHIVE_TOTAL_SIZE_LIMIT and ARCHIVE_FILE_SIZE_LIMIT parameters.
One is for the default of the first, another is for the default of the second.

Rather, any reasons that we have two parameters?

muga replied:
@sonots-zozo we may want to limit bytes of one file for `DEFAULT_ARCHIVE_FILE_SIZE_LIMIT` in the future instead of total size limit. The current `ARCHIVE_FILE_SIZE_LIMIT = ARCHIVE_TOTAL_SIZE_LIMIT` doesn't make sense but, 2 parameters are fine for me. What do you think?

@hiroyuki-sato Yes, `DEFAULT_ARCHIVE_FILE_SIZE_LIMIT` could be removed at least for now. What do you think?

hiroyuki-sato replied:
@sonots-zozo , @muga Thanks

* Remove `DEFAULT_ARCHIVE_FILE_SIZE_LIMIT`
* Change `DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT` order

sonots-zozo replied:
@muga Do you mean, for now, we can have only 

* api.max_archive_total_size_limit
* DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT

but we still have

* MAX_ARCHIVE_TOTAL_SIZE_LIMIT = systemConfig.get("api.max_archive_total_size_limit", Integer.class, DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT);
* MAX_ARCHIVE_FILE_SIZE_LIMIT = MAX_ARCHIVE_TOTAL_SIZE_LIMIT

and we may fix in the future?

It is fine for me as long as a comment is left at `MAX_ARCHIVE_FILE_SIZE_LIMIT = MAX_ARCHIVE_TOTAL_SIZE_LIMIT`.

hiroyuki-sato replied:
@sonots-zozo 
What do you think about this modification?

```diff
diff --git a/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java b/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java
index 5a1e20f76f..4b15b5634a 100644
--- a/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java
+++ b/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java
@@ -159,6 +159,7 @@ public class ProjectResource
     private static final Logger logger = LoggerFactory.getLogger(ProjectResource.class);
     private static int MAX_ARCHIVE_TOTAL_SIZE_LIMIT;
     private static final int DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT = 2 * 1024 * 1024;
+    // `MAX_ARCHIVE_FILE_SIZE_LIMIT = MAX_ARCHIVE_TOTAL_SIZE_LIMIT` for now
     private static int MAX_ARCHIVE_FILE_SIZE_LIMIT;
     private static int MAX_SESSIONS_PAGE_SIZE;
     private static final int DEFAULT_SESSIONS_PAGE_SIZE = 100;
```

muga replied:
@sonots-zozo sorry to your confusing. Yes, about `MAX_ARCHIVE_FILE_SIZE_LIMIT` (former `ARCHIVE_FILE_SIZE_LIMIT`), we want to fix it and make it configureable as well in the future to stop pushing project archive has unexpected files more earlier on digdag-server. I can leave a comments around `MAX_ARCHIVE_FILE_SIZE_LIMIT = MAX_ARCHIVE_TOTAL_SIZE_LIMIT`.

And to make the limit of archive total size configureable, these 3: `api.max_archive_total_size_limit`, `DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT`, `MAX_ARCHIVE_TOTAL_SIZE_LIMIT` are good. Users can specify the limit by `api.max_archive_total_size_limit`. If the system config is specified at bootstrap of Digdag, Digdag sets the value to `MAX_ARCHIVE_TOTAL_SIZE_LIMIT`. Otherwise, it uses `DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT` for the limit. I think that they are ok for us. Please correct me if wrong.

sonots-zozo replied:
@muga thanks, it is clear. It is fine for me for now.

@hiroyuki-sato 

```
// TODO: we may want to limit bytes of one file for `MAX_ARCHIVE_FILE_SIZE_LIMIT ` in the future instead of total size limit.
```

or muga-san can write about it.

hiroyuki-sato replied:
Fixed in 1edde99. Please take a look again.

muga commented:
@hiroyuki-sato about testing this feature, can we add an integration test by minimizing the threshold like `api.max_archive_total_size_limit=<smaller bytes size>` in digdag-tests?
https://github.com/treasure-data/digdag/blob/master/digdag-tests/src/test/java/acceptance/ServerJmxIT.java#L39

sonots-zozo commented on digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java:
```diff
@@ -204,6 +206,8 @@ public ProjectResource(
         this.projectArchiveLoader = projectArchiveLoader;
         MAX_SESSIONS_PAGE_SIZE = systemConfig.get("api.max_sessions_page_size", Integer.class, DEFAULT_SESSIONS_PAGE_SIZE);
+        MAX_ARCHIVE_TOTAL_SIZE_LIMIT = systemConfig.get("api.max_archive_total_size_limit", Integer.class, DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT);
+        MAX_ARCHIVE_FILE_SIZE_LIMIT = MAX_ARCHIVE_TOTAL_SIZE_LIMIT;
```
Should use `MAX_ARCHIVE_FILE_SIZE_LIMIT` as default value.

muga replied:
yes, that makes more sense.

sonots-zozo commented on digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java:
```diff
@@ -204,6 +205,8 @@ public ProjectResource(
         this.scsp = scsp;
         this.projectArchiveLoader = projectArchiveLoader;
         MAX_SESSIONS_PAGE_SIZE = systemConfig.get("api.max_sessions_page_size", Integer.class, DEFAULT_SESSIONS_PAGE_SIZE);
+        MAX_ARCHIVE_TOTAL_SIZE_LIMIT = systemConfig.get("api.max_archive_total_size_limit", Integer.class, DEFAULT_ARCHIVE_TOTAL_SIZE_LIMIT);
```
@hiroyuki-sato is it possible to print debug log to print the value of MAX_ARCHIVE_TOTAL_SIZE_LIMIT because, otherwise, I was not sure my configuration was correctly effective or not.

hiroyuki-sato replied:
I agree with you. but users may want to know another parameter, for example, `max_attempts_page_size`.
It is better to create another issue for logging.
And I have no idea how to output that parameter after launch digdag.

@muga , @yoyama Do you have any idea?

sonots-zozo replied:
Agree > users may want to know another parameter, for example, `max_attempts_page_size`.

Another PR is fine.

hiroyuki-sato commented:
@muga Thank you for your reply.
I just didn't know about that(The place of Integration test is in digdag-test).
Could you tell me what path should I use?

muga commented on digdag-server/build.gradle:
```diff
@@ -3,6 +3,7 @@ dependencies {
     compile project(':digdag-core')
     compile project(':digdag-client')
     compile project(':digdag-guice-rs-server-undertow')
+    testCompile project(path: ':digdag-tests', configuration: 'testArtifacts')
```
Please remove this. Not necessary any more.

muga commented on digdag-server/src/test/java/io/digdag/server/rs/ProjectResourceTest.java:
```diff
@@ -0,0 +1,65 @@
+import static utils.TestUtils.copyResource;
+import static utils.TestUtils.main;
+
+public class ProjectResourceTest
```
Please move this test class to `digdag-tests/src/test/java/` and rename package and class with `acceptance` and `LimitProjectArchiveFileSizeIT`. `IT` means integration tests in Digdag.

hiroyuki-sato commented:
@muga Thank you for your comment.
FIxed it and I moved the test into `digdag-test`.

Each test works as I expected. But when I run two tests at once.
I got the following error. I'm not yet sure how to solve it.

```
2019-02-22 15:35:18 +0900: Digdag v0.9.33
2019-02-22 15:35:18 +0900 [INFO] (pool-1-thread-2) io.digdag.core.crypto.SecretCryptoProvider: secret encryption engine: disabled
error: com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error injecting constructor, org.weakref.jmx.JmxException: io.digdag.core.workflow:name=TaskQueueDispatcher
  at org.weakref.jmx.guice.GuiceMBeanExporter.<init>(GuiceMBeanExporter.java:34)
  at org.weakref.jmx.guice.MBeanModule.configure(MBeanModule.java:38) (via modules: com.google.inject.util.Modules$OverrideModule -> io.digdag.guice.rs.server.jmx.JmxModule -> org.weakref.jmx.guice.MBeanModule)
  while locating org.weakref.jmx.guice.GuiceMBeanExporter

1 error (runtime)
```

muga commented:
@hiroyuki-sato thank you for adding tests! I fetched your change to my local and confirmed that your tests work somehow. Can you give me your command to execute the above integration tests? My command was:
```
$ CI_ACCEPTANCE_TEST=true ./gradlew digdag-tests:test --info --tests acceptance.LimitProjectArchiveFileSizeIT
```

hiroyuki-sato commented:
@muga I tested using IntelliJ.

* Test all class: `RunLimiteProjectArchiveFielSizeIT` -> above error occured.
* Test each methods: `runUploadProject` and `uploadProjectLargerThanLimit` -> passed

Anyway,  CI(2ecd94b) passed.

![Uploading スクリーンショット 2019-02-22 17.40.56.png…]()

muga commented:
@hiroyuki-sato our integration tests require environment variable `CI_ACCEPTANCE_TEST=true`. I'm not sure that you were configuring it during testing anyway.

Thank you for introducing this option. I merge this PR into master.

hiroyuki-sato commented:
Thank you!

hiroyuki-sato commented:
@muga Thank you!. I'll use `CI_ACCEPTANCE_TEST=true` at next time.
